### ISSUE TITLE
Karate upgrade from 1.3.1 to 1.4.0 not working correctly with Karate plugin 

### ISSUE THREAD

**Issue by Butipm** [karatelabs/karate-intellij-plugin#26]
Good day,

We fairly new to Karate and  recently upgraded from a working version 1.3.1 to 1.4.0, since cucumber-java is not supported anymore or at least no longer able to run test with it locally on Intellij (Build #IU-221.5787.30, built on May 31, 2022) with cucumber ,the recommended solution is to use the Karate plugin as per documentation , however, I am experiencing couple of issues as follow:

NB: We testing both UI, API tests, and some database validations tests with extended js scripts

_1.Performance issues with plug-in:_ 

Tests are not running consistently when debugging, either they hang at "initialization",  I then have to re-run them which it takes time to load then fails or they run and pass which I am not sure at what condition.
NB: I also noticed I had to select "Shorten the test name on command line"

2. Code is not highlighted as it used to be before on 1.3.1:

![image](https://user-images.githubusercontent.com/44231371/233996862-c31865e9-32e1-4e38-9aa6-d3aa792e482f.png)

3. Test no longer run with chromedriver:
driver config on Karate-config file:
type: 'chromedriver',
			executable: 'src/test/java/chromedriver.exe',
			addOptions: ['--ignore-certificate-errors'],
			headless: false

[karate-config.js.txt](https://github.com/karatelabs/karate-intellij-plugin/files/11310469/karate-config.js.txt)

NB: we had to use chromedriver temporarily after chrome upgrade broke tests, but I believe with 1.4.0 it has been resolved

4. Remote test to grid no longer run if we change the environment variable on karate-config.js:
Test run locally instead (if they run at all). see attached for karate-config.js file fo

NB: We need these tests to run remotely (In seleniumbox/Grid) when we run our code build pipeline

[karate-config.js.txt](https://github.com/karatelabs/karate-intellij-plugin/files/11310469/karate-config.js.txt)

POM
[pom.txt](https://github.com/karatelabs/karate-intellij-plugin/files/11310550/pom.txt)

Runner class
[runner class.txt](https://github.com/karatelabs/karate-intellij-plugin/files/11310549/runner.class.txt)

Run config:

![image](https://user-images.githubusercontent.com/44231371/234001030-f4a8b79d-cebc-4c19-8178-f65addd5e202.png)

**Comment by Butipm** [comment]
OK, I will follow the above the instrauctions. In the meantime to replace the issue please find attached POM, Karate-config above on the issue

**Comment by Butipm** [comment]
[test-automation.zip](https://github.com/karatelabs/karate-intellij-plugin/files/11321494/test-automation.zip)
Good day Mr. Peter, 

Please find attached a working project to be upgraded Karate version 1.4.0 with IntelliJ IDEA 2022.1.2 (Build #IU-221.5787.30, built on May 31, 2022) with Karate Plug-in

**Comment by Butipm** [comment]
Hi @ptrthomas , I think I may have resolved the issue but I still need your advise here.

1. 
- uninstalled Intellij IDEA completely and reinstalled  it (Build #IU-231.8770.65, built on April 27, 2023)
-  only installed Karate plugin
- switched from typedriver: chromedriver to native chrome (NB: it would be great if chromedriver Issue may be resolved as this may be a workaround should we have another chrome upgrade that blocks karate like in the case we recently had with version 1.3.1)

Then the performance of the plugin improved and now I could execute tests (with minor changes to the syntax especially on scenario outlines)

2. Scenario outlines with multiple rows of data seems to fail on the 2nd test iteration ,driver picks a different port on the browser:
The problem is the url is not loaded by the driver on the 2nd (sometimes on the first iteration), as a result the test fails trying to locate web elements which are obviously not there.

To work around this issue I implemented a check  to validate the url is loaded by the driver if not load it again: example below.
-----------------------------------------
	         Given driver webUrl
		* def actualUrl = driver.url
		* eval if (actualUrl === '') (driver.url = webUrl)
----------------------------------------

This checks if the driver loaded the the url if not it reloads it again.

NB: I also tried to close the driver at the end of the test (* driver.close()/  * driver.quit() ) but that did not help

If this is sufficient please advice to close it

**Comment by Butipm** [comment]
furthermore not sure I should log a different issue here but our selenium grid capabilities are now failing, they work fine on version 1.3.1 but on 1.4.1 they fail with this error : {"status":13,"value":{"error":"13","message":"Session [(null externalkey)] not available and is not among the last 1000 terminated sessions

------------------------------------------------------
function fn() {
	// get system property from maven cmd eg 'karate.env'
	var env = karate.env;
	var testWeburl = karate.properties['testWeburl'];
	var testAPIurl = karate.properties['testAPIurl'];
	var browser = karate.properties['browser'];
	var testUser = karate.properties['testUser'];
	var testPass = karate.properties['testPass'];
	var selniumBToken = karate.properties['selniumBToken'];
	karate.log('karate.env system property was:', env);

	// Set defaults in case of running local environment, web and api Url
	if (!env) {
		env = 'remote';

		testWeburl = 'https://url.com/';


		testAPIurl = 'https://api.url.co/';
		selniumBToken = '8649cjebjbefj-ebs';
// set your own

		var testUser = java.lang.System.getenv("UserName");
		var testPass = java.lang.System.getenv("PassWord");
	

	}
	// Set defaults in case of running local: Browser and driver as per local and selenium-box
	if (!browser) {
		browser = 'chrome';
		typedriver = 'chrome';
	} else if (browser === 'chrome') {
		browser = 'chrome';
		typedriver = 'chromedriver';
	} else if (browser === 'firefox') {
		browser = 'firefox';
		typedriver = 'geckodriver';
	} else if (browser === 'MicrosoftEdge') {
		browser = 'MicrosoftEdge';
		typedriver = 'msedgedriver';
	} else if (browser === 'iexplore') {
		browser = 'internet explorer';
		typedriver = 'iedriver';
	}

// assign the variable either from Mvn cmd or defaults for driver config
	var config = {
		env: env,
		testWeb: testWeburl.toString(),
		testUserName: testUser.toString(),
		testPassWord: testPass.toString(),
		testNewSubscriptionUrl: `${testWeburl}/create`,
		testApi: testAPIurl.toString(),
		browsertype: browser.toString(),
	}
	var retryCount = 3; // set the number of times to retry
	var retryDelay = 5000; // set the delay between retries in milliseconds

	function configureDriverWithRetry(driverConfig, retryCount) {
		var retry = retryCount || 1;
		try {
			karate.configure('driver', driverConfig);
		} catch (error) {
			if (retry > 0) {
				karate.log('Error configuring driver, retrying ...');
				karate.log(error);
				karate.sleep(5000);
				configureDriverWithRetry(driverConfig, retry - 1);
			} else {
				karate.log('Failed to configure driver after retries');
				// Log error stack trace for detailed error information
				karate.log(error.stack);
				throw error;
			}
		}
	}

	if (env === 'local') {

		// run test on the local machine, note if you are using 32-bit machine , change the executable path 'C:/Program Files (x86)/.../chrome'
		var driverConfig1 = {
			type: 'chrome',
			executable: 'C:/Program Files/Google/Chrome/Application/chrome.exe',
			addOptions: ['--ignore-certificate-errors'],
			port:9222,
			headless: false
		};
		karate.configure('driver', driverConfig1);


	} else if (env === 'remote') {

		//For Remote execution use the below seleniumbox capabilities
		var session = {
			capabilities: {
				browserName: browser
			},
			desiredCapabilities: {
				"e34": true,
				"e34:token": selniumBToken,
				"e34:video": true,
				"takesScreenshot": true,
				browserName: browser,
				"e34:pacAlias": 'proxy.com',
				"e34:version": "110.0.5481.77-1",
				"e34:l_testName": karate.info.scenarioName
			}
		};
		var driverConfig = {
			type: typedriver,
			addOptions: ['--ignore-certificate-errors'],
			start: false,
			webDriverUrl: 'https://seleniumbox.net/wd/hub',
			webDriverSession: session
		}
		configureDriverWithRetry(driverConfig, 4); // retry 3 times

	}
	//load weblocators from json file
	var result = read('classpath:locators.json')
	config.webLoc = result;

	karate.configure('connectTimeout', 60000);
	karate.configure('readTimeout', 60000);
	karate.configure('logPrettyRequest', true);
	karate.configure('logPrettyResponse', true);
	karate.configure('retry', {count: 10, interval: 1500});


	return config;
}

**Comment by ptrthomas** [comment]
@Butipm - for the chrome / selenium issue I would like to create a separate issue at the karate project itself, and following this process: https://github.com/karatelabs/karate/wiki/How-to-Submit-an-Issue - and I would greatly appreciate if you can simplify your project (for example karate-config.js is too complex)

can I close this issue - because I don't think this is related to the intellij plugin. to double-check, see if you get the same problems when you run from the command-line

browser automation on the local machine with chromedriver is not easy - and will not work properly when you start running in parallel, this is why teams use remote servers or docker / cloud options: https://stackoverflow.com/a/62325328/143475

**Comment by Butipm** [comment]
Hi @ptrthomas ,

Understood, I was able to run the test with the same karate-config using the cmd below
mvn clean test -DselniumBToken=8649c5aadvhjv -Dkarate.env=remote -DuiWeburl=https://ui.co -DAPIurl=https://api.com -Dbrowser=chrome -DiwaUser=<userName> -DiwaPass=<password> -Dkarate.options="--tags @end2endTests" -f pom.xml. Please go ahead and close the issue , I will do more investigation on my side and log another issue  for this one
